OpenChange/RTFHandler.m

```diff
@@ -529,109 +763,137 @@ - (RTFColorTable *) parseColorTable
-                                                                length: len-1
-                                                              encoding: NSASCIIStringEncoding
-                                                          freeWhenDone: NO];
-
-              cw = [self parseControlWord: &len];
-              if (len == 0)  // Possibly parsing a space
-                cw = [self parseControlWord: &len];
-
-              fontInfo->charset = [[NSString alloc] initWithBytesNoCopy: (void *)cw+1
-                                                                 length: len-1
-                                                               encoding: NSASCIIStringEncoding
-                                                           freeWhenDone: NO];
-
-              // We now skip everything until we find our final group closer ('}')
-              int cc = 1;
-              
-              do
-                {
-                  if (*_bytes == '{')
-                    cc++;
-                  if (*_bytes == '}')
-                    cc--;
-                  
-                  ADVANCE;
-                }
-              while (cc != 0);
-
-              // move back our buffer;
-              REWIND;
+            case FONTCW_F: 
+              if (hasArg)
+                fontInfo->index = arg;
+              break;
+            case FONTCW_FBIDI:
+              fontInfo->family = @"bidi";
+              break;
+            case FONTCW_FCHARSET: 
+              if (hasArg) 
+                fontInfo->charset = arg;
+              break;
+            case FONTCW_FDECOR:
+              fontInfo->family = @"decor";
+              break;
+            case FONTCW_FMODERN:
+              fontInfo->family = @"modern";
+              break;
+            case FONTCW_FNIL: 
+              fontInfo->family = @"nil"; 
+              break;
+            case FONTCW_FPRQ:
+              if (hasArg)
+                fontInfo->pitch = arg;
+              break;
+            case FONTCW_FROMAN:
+              fontInfo->family = @"roman"; 
+              break;
+            case FONTCW_FSCRIPT:
+              fontInfo->family = @"script";
+              break;
+            case FONTCW_FSWISS:
+              fontInfo->family = @"swiss";
+              break;
+            case FONTCW_FTECH:
+              fontInfo->family = @"tech";
+              break;
+            case FONTCW_UNKNOWN:
+            default:
+              // do nothing
+              break;
             }
         }
-      else
+      else // no char
         {
-          if (isalnum(*_bytes))
-            [fontName appendFormat: @"%c", *_bytes]; 
+          if (level == 2 && isalnum(*_bytes))
+            {
+              // we assume this is the fontname
+              // there is no limit length but we took 100
+              // as protection
+              const char *delim = strpbrk(_bytes, ";{}\\");
+              unsigned int fontnameLen = delim - _bytes;
```

Comment:
That was a good catch, moreover the rtf should be very bad formed to delim == NULL so it will not be detected by the tests.